### PR TITLE
Extend with SVM probes

### PR DESCRIPTION
**Pull request by rikardvinge**
This PR implements polynomial kernel SVM probing, refactors probing code to simplify future probe extensions and some qualilty of life updates for the repository.

- Refactor probing for simplified extension: Probes now reside in their own subfolder and the two current probes, single-layer perceptron and poly kernel SVM have been given the same interface.
- Polynomial kernel SVM probe: Uses Bayesian hyperparameter search through Scikit-Optimize to tune hyperparameters for the probe. HP parameter search is not yet implemented for single-layer perceptron. The SVM allows for linear probing as well as non-linear probing with complete control over the degree of nonlinearity. This is useful for analyzes into how nonlinear do heads need to be for a specific task.
- Updated documentation for the new probing structure.
- Improved GH workflows.

### CODE REVIEW

File: benchmark/evaluation/visualisations.py [hunk 1]
```diff
@@ -80,28 +81,35 @@ def plot_regression_scatter(
 
     # Plot training and validation points
     ax.scatter(y_train_true, y_train_pred, alpha=0.3, label="Train")
-    ax.scatter(y_val_true, y_val_pred, alpha=0.6, label="Validation")
+
+    if (y_val_true is not None) and (y_val_pred is not None):
+        ax.scatter(y_val_true, y_val_pred, alpha=0.6, label="Validation")
+        bounds = [min([min(y_train_true), min(y_val_true)]),
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/visualisations.py]: you could skip the nested list here and simplify to `[min(min(y_train_true), min(y_val_true)),`

File: benchmark/evaluation/probes/linear.py [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+        self.enable_plots = enable_plots
+        self.embedding_dim = embedding_dim
+        self.probe_params = probe_params
+        self.splitter = splitter
+        self.store_models = store_models
+        
+        self.dataset = EmbeddingDataset(df)
+        self.trainer = None
+        
+        self.batch_size = self.probe_params.pop('batch_size', 64)
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/linear.py]: I would prefer to not have "silent" defaults for batch size, lr and epochs as they are quite influential on the results and we did not fix them for the benchmark. Can we make them required again? If you disagree, lets at least add an informative warning "Batch size not set, defaulting to .."

File: docs/evaluation_and_results.md [hunk 3]
```diff
@@ -158,3 +169,28 @@ task_filter:
 standardize_embeddings: false
 normalize_labels: false
 ```
+
+### Working with stored models
```
Comment (Isabelle-Wittmann) [inline docs/evaluation_and_results.md]: Just as an idea: Do you think the use of model could be ambiguous in our repo as it can refer to an encoder/ embedding model, and the probing model? We could also always refer to the NeuCo-fitted model as Probe? But totally we fine with keeping it as you have it now :)

File: configs/sample_config_linear.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+# Optional configs
+embedding_dim: null
+standardize_embeddings: true
+normalize_labels: true
+enable_plots: false
+update_leaderboard: false
+store_models: false
+task_filter: null #eg. ["biomass_mean", "biomass_std"]
+probe_type: "linear"
+probe_params:
```
Comment (Isabelle-Wittmann) [inline configs/sample_config_linear.yaml]: As said above, I would prefer to have some probe_params as required

File: configs/sample_config_linear.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+# Optional configs
+embedding_dim: null
+standardize_embeddings: true
+normalize_labels: true
+enable_plots: false
+update_leaderboard: false
+store_models: false
+task_filter: null #eg. ["biomass_mean", "biomass_std"]
+probe_type: "linear"
```
Comment (Isabelle-Wittmann) [inline configs/sample_config_linear.yaml]: We could also make probe_type required if you think that makes it more transparent what is used

File: benchmark/evaluation/utils.py [hunk 6]
```diff
@@ -13,4 +17,23 @@ def fix_all_seeds(seed: int = 42):
...
+
+
+@dataclass
+class FoldResult:
+    train_loss: list[float]
+    val_loss: list[float]
+    metric_history: list[float]
+    best_metric: float
+    best_model_state: dict
+    model: Optional[Any] = None
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/utils.py]: Might not have used this consequently myself but generally` Any | None` is the more modern syntax, and I'd rather use that over `Optional`

File: benchmark/evaluation/probes/linear.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+            learning_rate: Learning rate for Linear Probe training.
+            fold_index: Index of split. Used in plotting.
+            output_dir: Path to folder to store results in.
+            filename_prefix: Prefix to add to files of visualizations.
+            enable_plots: Flag controlling the creation of plots. Set to True to plot results.
+            epochs: Default number of epochs to train network.
+        """
+
+        self.init_params = {'input_dim': embedding_dim}
+        self.model = LinearProbe(**self.init_params)
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/linear.py]: I think the `LinearProbe` model never gets moved to device before training, can you double-check that?

File: benchmark/evaluation/probes/linear.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+                self.optimizer.step()
+        return loss.detach(), preds.detach()
+    
+    def fit_only(self, 
+            data_loader: DataLoader,
+            epochs: int = None,
+            refit: bool = False,
+            ):
+        if refit:
+            self.model = LinearProbe(**self.init_params)
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/linear.py:125]: If you init a new model here, you also need to reinit the optimizer and point it to the new model params `self.optimizer = torch.optim.AdamW(self.model.parameters(), lr=learning_rate)`

Comment (Isabelle-Wittmann) [reply benchmark/evaluation/probes/linear.py:125]: for that you also need to store the learning rate in self.learning_rate or similar

File: benchmark/evaluation/probes/linear.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+               ):
+        
+        self.task_type = task_type
+        self.task_name = task_name
+        self.device = device
+        self.output_dir = output_dir
+        self.filename_prefix = filename_prefix
+        self.enable_plots = enable_plots
+        self.embedding_dim = embedding_dim
+        self.probe_params = probe_params
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/linear.py]: With this, self.probe_params references the original `probe_params` dict, so pop("batch_size") also modifies the external object. I'd use `self.probe_params = dict(probe_params) `to create a copy instead.

File: benchmark/evaluation/probes/linear.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+
+        for _ in range(epochs):
+            # Training phase
+            self.model.train()
+            epoch_train_losses, train_preds, train_targets = [], [], []
+            for batch in train_loader:
+                loss, preds = self._step(batch, training=True)
+                epoch_train_losses.append(loss)
+                train_preds.append(preds)
+                train_targets.append(batch[1])
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/linear.py]: for gpu training: batch is still not on device while preds are (after fixing the other comment that moves the model to device), so either you need to move batch to device here, or otherwise you could return targets from the _step function which is already on device

File: benchmark/evaluation/probes/svm.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+    def __init__(
+            self,
+            task_type: str,
+            task_name: str,
+            output_dir: Path,
+            filename_prefix: str,
+            splitter: Callable, 
+            enable_plots: bool, 
+            df: pd.DataFrame = None,
+            C_start_end: Optional[list[float]] = None,
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/svm.py]: As mentioned somewhere else, for more modern syntax I'd switch to `list[float] | None`

File: benchmark/evaluation/probes/svm.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+            self.param_space['gamma'] = Real(kernel_gamma_start_end[0], kernel_gamma_start_end[1], prior='log-uniform')
+
+        self.opt = BayesSearchCV(
+            model,
+            self.param_space,
+            return_train_score=True,
+            scoring={'metric': self.metric_fn, 'scoring': self.scoring_fn},
+            cv=splitter,
+            refit='scoring',
+            **opt_params,
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/svm.py:193]: opt_params is optional and unpacking will throw an error if opt_params is None

File: benchmark/evaluation/probes/linear.py [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+                filename_prefix: str,
+                enable_plots: bool,
+                learning_rate: float = 0.001,
+                epochs: int = 10,
+            ) -> None:
+        """Train and evaluate a single fold.
+        
+        Args:
+            model: torch.Model to use in evaluation.
+            task_type: Type of task, either "classification" or "regression".
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/linear.py]: tiny detail, but lets also update this to include cls and regr

File: benchmark/evaluation/probes/linear.py [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+            fold_index: Index of split. Used in plotting.
+            output_dir: Path to folder to store results in.
+            filename_prefix: Prefix to add to files of visualizations.
+            enable_plots: Flag controlling the creation of plots. Set to True to plot results.
+            epochs: Default number of epochs to train network.
+        """
+
+        self.init_params = {'input_dim': embedding_dim}
+        self.model = LinearProbe(**self.init_params)
+        self.task_type = task_type
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/linear.py]: this should also be task_type.lower() to be consistent with svm

File: benchmark/evaluation/probes/svm.py [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+            splitter (Callable): Splitter used for validation statistics, e.g. sklearn.model_selection.ShuffleSplit.
+            enable_plots (bool): Whether to generate plots or not. Default is False.
+            df (pd.DataFrame, optional): Dataframe to use for training and validation.
+            C_start_end (list[float], optional): List of start and end values for C hyperparameter. Defaults to None.
+            kernel_degree_list (list[int], optional): List of kernel degrees to evaluate. Defaults to [1]. Note that `kernel_gamma_start_end` cannot be used if 1 is in `kernel_degree list`.
+            kernel_gamma_start_end (list[float], optional): List of start and end values for gamma hyperparameter. Defaults to [1e-6, 10]. Note that `kernel_gamma_start_end` cannot be used if 1 is in `kernel_degree list`.
+            opt_params (dict[str, Any], optional): Keyword input arguments for skopt.BayesSearchCV.
+        """
+
+        self.task_type = task_type
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/svm.py]: I would do task_type.lower() here and then reuse self.task_type below for checking cls/regr

File: benchmark/evaluation/probes/svm.py [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+            self.x = np.stack(df['embedding'].to_numpy())
+            self.y = np.stack(df['label'].to_numpy())
+        else:
+            self.x = None
+            self.y = None
+
+        if task_type.lower() in ("classification", "cls"):
+            self.model_type = SVC
+            self.metric_fn = 'f1'
+            self.scoring_fn = 'neg_log_loss'
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/svm.py]: for a log loss you need returned probabilities, did you test classification/ does the model return probabilities here? You might have to set SVC(probability=True) for this as far as I know, but haven't tested it

File: benchmark/evaluation/probes/svm.py [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+        """
+        Fit the Bayesian hyperparameter optimization search on the training data.
+
+        Args:
+            x (np.ndarray, optional): Training features of shape (n_samples, n_features).
+            y (np.ndarray, optional): Training targets of shape (n_samples,).
+        """
+        if (x is None) and (y is None):
+            x = self.x
+            y = self.y
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/svm.py:240]: Dont you need another None check after this as self.x and self.y can be None as well and fit will fail then?

File: benchmark/evaluation/probes/svm.py [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+            scoring={'metric': self.metric_fn, 'scoring': self.scoring_fn},
+            cv=splitter,
+            refit='scoring',
+            **opt_params,
+        )
+
+    def fit_from_state(self, state, x=None, y=None):
+        if (x is None) and (y is None):
+            x = self.x
+            y = self.y
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/svm.py:215]: See comment below R130

File: benchmark/evaluation/probes/svm.py [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+        """Save the fitted SVM model and write a Python script to load it.
+
+        Args:
+            path: Directory path where model files will be saved.
+        """
+        path = Path(path)
+        path.mkdir(parents=True, exist_ok=True)
+
+        if self.model is None:
+            raise RuntimeError("Model has not been fitted yet. Call optimize_score() first.")
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/svm.py]: should this be Call train() first?

File: benchmark/evaluation/visualisations.py [hunk 20]
```diff
@@ -138,7 +147,8 @@ def plot_confusion_matrix(
     ax.set_ylabel("True label")
     
     # Save figure
-    filename = f"{base_name}_{task_name}_cm_fold{fold_idx + 1}.png"
+    fold_save_str = f"_fold{fold_idx + 1}"
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/visualisations.py]: Add fold_idx = None condition

File: benchmark/evaluation/visualisations.py [hunk 21]
```diff
@@ -80,28 +81,35 @@ def plot_regression_scatter(
...
+        bounds = [min([min(y_train_true), min(y_val_true)]), 
+                  max([max(y_train_true), max(y_val_true)])]
+    else:
+        bounds = [min(y_train_true), max(y_train_true)]
 
     # Draw diagonal identity line
-    bounds = [min(min(y_train_true), min(y_val_true)), max(max(y_train_true), max(y_val_true))]
     ax.plot(bounds, bounds, linestyle='--', linewidth=3, color='green')
 
+    fold_plot_str = ", Fold {fold_idx + 1}" if fold_idx is not None else ""
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/visualisations.py]: Missing f-string

File: benchmark/evaluation/linear_probing.py
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/linear_probing.py:1]: Can we rename this to something more general now, probing/ cross-validation/ ...

File: benchmark/evaluation/linear_probing.py [hunk 22]
```diff
@@ -277,12 +130,19 @@ def cross_validate(
         "q_stat": q_stat,
         "mean_score": mean_score,
         "std_dev": std_dev,
+        "hyperparamters": best_params,
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/linear_probing.py]: typo hyperparameters

File: benchmark/evaluation/probes/svm.py [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+
+        self.task_type = task_type
+        self.task_name = task_name
+        self.output_dir = output_dir
+        self.filename_prefix = filename_prefix
+        self.enable_plots = enable_plots
+        self.splitter = splitter
+        self.model = None
+        self.best_state = None
+        self.is_optimized = False
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/svm.py]: what is self.is_optimized needed/used for?

**Review by Isabelle-Wittmann — COMMENTED** [review]
Hey, thanks for the PR, generally looks good and much cleaner! I've added some smaller/ optional comments, but also some comments related to GPU training (device calls) and potentially edge case failures if certain variables are None that should be addressed before I approve. I did not play around with running the code in various settings yet, and will do that once comments are addressed for a final check :) THX!

File: benchmark/evaluation/probes/svm.py (hunk 23 above)
Comment (rikardvinge) [reply benchmark/evaluation/probes/svm.py]: Nothing, leftover from previous design. I'll remove it

File: benchmark/evaluation/linear_probing.py (hunk 22 above)
Comment (rikardvinge) [reply benchmark/evaluation/linear_probing.py]: Fixed

File: benchmark/evaluation/visualisations.py (hunk 20 above)
Comment (rikardvinge) [reply benchmark/evaluation/visualisations.py]: Fixed

File: benchmark/evaluation/visualisations.py (hunk 21 above)
Comment (rikardvinge) [reply benchmark/evaluation/visualisations.py]: Fixed

File: benchmark/evaluation/probes/linear.py (hunk 2 above)
Comment (rikardvinge) [reply benchmark/evaluation/probes/linear.py]: I made both probe_type and probe_params to required, and added some extra config checks to ensure the different models get what they need.

File: benchmark/evaluation/visualisations.py (hunk 1 above)
Comment (rikardvinge) [reply benchmark/evaluation/visualisations.py]: Changed

File: docs/evaluation_and_results.md (hunk 3 above)
Comment (rikardvinge) [reply docs/evaluation_and_results.md]: I renamed all references where we store the trained probe from "model" to "probe". Good idea!

File: benchmark/evaluation/probes/svm.py (hunk 18 above)
Comment (Isabelle-Wittmann) [reply benchmark/evaluation/probes/svm.py:215]: See https://github.com/embed2scale/NeuCo-Bench/pull/12#discussion_r3199873880

File: benchmark/evaluation/probes/svm.py [hunk 24]
```diff
@@ -40,6 +44,30 @@ def validate_config(probe_params: dict):
     assert len(probe_params['C_start_end']) == 2, f"`C_start_end` must have exactly two elements."
     assert probe_params['C_start_end'][0] < probe_params['C_start_end'][1], f"`C_start_end` must be formatted [minimum C, maximum C] but got {probe_params['C_start_end']}."
 
+    # epsilon checks
+    epsilon_list = probe_params.get('epsilon_list', None)
+    if epsilon_list is not None:
+        assert isinstance(epsilon_list, (list, float)), f"`epsilon_list` must be float or list but got {type(epsilon_list)}."
```
Comment (Isabelle-Wittmann) [inline benchmark/evaluation/probes/svm.py:50]: does it make sense to add `int`?

**Comment by Isabelle-Wittmann** [comment]
Hi, thx again all looks good and ran well. Please also update https://github.com/embed2scale/NeuCo-Bench/blob/main/docs/evaluation_and_results.md in this or a subsequent PR for the updated config structure that is required, maybe we should also add somewhere that this PR breaks backward compatibility for the config format?
